### PR TITLE
[FIX] base: handle ValueError for invalid domain

### PR DESCRIPTION
**Pull request by mobe-odoo**
This traceback appears when user adds a new field
(`many2one`, `one2many`, `many2many`) using studio and adds a invalid 
domain in `Domain` field.

Steps to reproduce:
1) Install `stock`, `web_studio`.
2) Open `Inventory` module > `Products` > `Products`. 
3) Click on any product to open its form view > `Studio` button. 
4) Drag and drop new `many2one` field.
5) Select any model (e.g `Product`) > `Confirm`.
6) Now click on that `New Many2One` field > `MORE` under `Properties`. 
7) Now enter any invalid domain (e.g `'qty_available'>10`) under `Properties`.

See the traceback:

```
TypeError: '>' not supported between instances of 'str' and 'int'
  File "odoo/tools/safe_eval.py", line 362, in safe_eval
    return unsafe_eval(c, globals_dict, locals_dict)
  ?, line 1, in <module>
ValueError: <class 'TypeError'>: "'>' not supported between instances of 'str' and 'int'" while evaluating
"['qty_available'>10]"
  File "odoo/http.py", line 2134, in __call__
    response = request._serve_db()
  File "odoo/http.py", line 1710, in _serve_db
    return service_model.retrying(self._serve_ir_http, self.env)
  File "odoo/service/model.py", line 133, in retrying
    result = func()
  File "odoo/http.py", line 1737, in _serve_ir_http
    response = self.dispatcher.dispatch(rule.endpoint, args)
  File "odoo/http.py", line 1938, in dispatch
    result = self.request.registry['ir.http']._dispatch(endpoint)
  File "odoo/addons/base/models/ir_http.py", line 191, in _dispatch
    result = endpoint(**request.params)
  File "odoo/http.py", line 717, in route_wrapper
    result = endpoint(self, *args, **params_ok)
  File "addons/web/controllers/dataset.py", line 30, in call_kw
    return self._call_kw(model, method, args, kwargs)
  File "addons/web/controllers/dataset.py", line 26, in _call_kw
    return call_kw(request.env[model], method, args, kwargs)
  File "odoo/api.py", line 461, in call_kw
    result = _call_kw_multi(method, model, args, kwargs)
  File "odoo/api.py", line 448, in _call_kw_multi
    result = method(recs, *args, **kwargs)
  File "home/odoo/src/enterprise/saas-16.4/web_studio/models/studio_mixin.py", line 33, in write
    res = super(StudioMixin, self).write(vals)
  File "odoo/addons/base/models/ir_model.py", line 1009, in write
    res = super(IrModelFields, self).write(vals)
  File "odoo/models.py", line 4069, in write
    real_recs._validate_fields(vals, inverse_fields)
  File "odoo/models.py", line 1439, in _validate_fields
    check(self)
  File "odoo/addons/base/models/ir_model.py", line 597, in _check_domain
    safe_eval(field.domain or '[]')
  File "odoo/tools/safe_eval.py", line 376, in safe_eval
    raise ValueError('%s: "%s" while evaluating\n%r' % (ustr(type(e)), ustr(e), expr))
```

When user enters this wrong domain, the ValueError is raised from `safe_eval` 
method. So to handle I have used try-except block.

sentry-4399811257

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Comment by mobe-odoo** [comment]
Hello @abz-odoo  
can you please check this PR?

**Comment by maan-odoo** [comment]
Hello @abz-odoo 
Please look it.
Thanks :slightly_smiling_face:

**Review by abz-odoo — CHANGES_REQUESTED** [review]

File: odoo/addons/base/models/ir_model.py [hunk 1]
```diff
@@ -594,7 +594,10 @@ def _in_modules(self):
     @api.constrains('domain')
     def _check_domain(self):
         for field in self:
-            safe_eval(field.domain or '[]')
+            try:
+                safe_eval(field.domain or '[]')
+            except ValueError:
+                raise ValidationError(_("Invalid domain %s" %field.domain))
```
[the remark this answers is not in the document]
Comment (mart-e) [reply odoo/addons/base/models/ir_model.py]: @abz-odoo not yet ;-) https://www.odoo.com/documentation/16.0/developer/howtos/translations.html#variables

**Comment by abz-odoo** [comment]
> @abz-odoo not yet ;-) https://www.odoo.com/documentation/16.0/developer/howtos/translations.html#variables


@maan-odoo can you apply this ?

File: odoo/addons/base/models/ir_model.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (maan-odoo) [reply odoo/addons/base/models/ir_model.py]: Done

**Comment by maan-odoo** [comment]
> > @abz-odoo not yet ;-) https://www.odoo.com/documentation/16.0/developer/howtos/translations.html#variables
> 
> @maan-odoo can you apply this ?

Done

**Review by xmo-odoo — COMMENTED** [review]
@robodoo override=ci/security

File: odoo/addons/base/models/ir_model.py [hunk 2]
```diff
@@ -597,7 +597,10 @@ def _in_modules(self):
     @api.constrains('domain')
     def _check_domain(self):
         for field in self:
-            safe_eval(field.domain or '[]')
+            try:
+                safe_eval(field.domain or '[]')
+            except ValueError:
+                raise ValidationError(_("Invalid domain %s", field.domain))
```
[the remark this answers is not in the document]
Comment (maan-odoo) [reply odoo/addons/base/models/ir_model.py]: Thanks for suggestion.  
I have updated the code.

File: odoo/addons/base/models/ir_model.py [hunk 3]
```diff
@@ -597,7 +597,10 @@ def _in_modules(self):
     @api.constrains('domain')
     def _check_domain(self):
         for field in self:
-            safe_eval(field.domain or '[]')
+            try:
+                safe_eval(field.domain or '[]')
+            except ValueError:
```
Comment (ryv-odoo) [inline odoo/addons/base/models/ir_model.py:602]: Actually, `save_eval` can crash for many other reasons than a `ValueError`. By example, if someone enter `5 +`, you will get a `SyntaxError`, but it should also catch this kind of `Exception`.
```suggestion
            except Exception:
```

Also, the message is ambiguous, you don't check the validity of the domain but only the Python code execution.

Note that it doesn't make any sense to have use `safe_eval` here. Using `literal_eval` should be enough. I will create a small PR in master for that.

Comment (maan-odoo) [reply odoo/addons/base/models/ir_model.py]: Hello @ryv-odoo 

Changes are done.
Please look at it.
Thank you!

File: odoo/addons/base/models/ir_model.py [hunk 4]
```diff
@@ -597,7 +597,12 @@ def _in_modules(self):
     @api.constrains('domain')
     def _check_domain(self):
         for field in self:
-            safe_eval(field.domain or '[]')
+            try:
+                safe_eval(field.domain or '[]')
+            except Exception:
+                raise ValidationError(
+                    _("Invalid Python code '%s' for the domain field.", field.domain)
+                ) from None
```
Comment (ryv-odoo) [inline odoo/addons/base/models/ir_model.py:605]: Can you add the reason for the failure ? It's still a technical view, the exception message can be very helpful.

```suggestion
            except Exception as e:
                raise ValidationError(
                    _("Invalid Python code '%s' for the domain field: %s", field.domain, e)
                ) from None
```

**Pull request closed without merging**